pySDC/helpers/spectral_helper.py

```diff
@@ -0,0 +1,1692 @@
+        Consider `put_BCs_in_rhs_hat` to add BCs with no extra transforms needed.
+
+        Args:
+            rhs: Right hand side in physical space
+
+        Returns:
+            rhs in physical space with BCs
+        """
+        assert rhs.ndim > 1, 'rhs must not be flattened here!'
+
+        ndim = self.ndim
+
+        for axis in range(ndim):
+            _rhs_hat = self.transform(rhs, axes=(axis - ndim,))
+
+            for bc in self.full_BCs:
+                slices = (
+                    [slice(0, self.init[0][i + 1]) for i in range(axis)]
+                    + [bc['line']]
+                    + [slice(0, self.init[0][i + 1]) for i in range(axis + 1, len(self.axes))]
+                )
+                if axis == bc['axis']:
+                    _slice = [self.index(bc['equation'])] + slices
+
+                    N = self.axes[axis].N
+                    if (N + bc['line']) % N in self.xp.arange(N)[self.local_slice[axis]]:
+                        _slice[axis + 1] -= self.local_slice[axis].start
+
+                        _rhs_hat[(*_slice,)] = bc['v']
+
+            rhs = self.itransform(_rhs_hat, axes=(axis - ndim,))
+
+        return rhs
+
+    def add_equation_lhs(self, A, equation, relations):
+        """
+        Add the left hand part (that you want to solve implicitly) of an equation to a list of lists of sparse matrices
+        that you will convert to an operator later.
+        For instance, the `relations` for diffusion in component `u` would look like `relations = {'u': -Dxx}`.
+
+        Args:
+            A (list of lists of sparse matrices): The operator to be
+            equation (str): The equation of the component you want this in
+            relations: (dict): Relations between quantities
+        """
+        for k, v in relations.items():
+            A[self.index(equation)][self.index(k)] = v
+
+    def convert_operator_matrix_to_operator(self, M):
+        """
+        Promote the list of lists of sparse matrices to a single sparse matrix that can be used as linear operator.
+
+        Args:
+            M (list of lists of sparse matrices): The operator to be
+
+        Returns:
+            sparse linear operator
+        """
+        if len(self.components) == 1:
+            return M[0][0]
+        else:
+            return self.sparse_lib.bmat(M, format='csc')
+
+    def get_wavenumbers(self):
+        """
+        Get grid in spectral space
+        """
+        grids = [self.axes[i].get_wavenumbers()[self.local_slice[i]] for i in range(len(self.axes))][::-1]
+        return self.xp.meshgrid(*grids)
+
+    def get_grid(self):
+        """
+        Get grid in physical space
+        """
+        grids = [self.axes[i].get_1dgrid()[self.local_slice[i]] for i in range(len(self.axes))][::-1]
+        return self.xp.meshgrid(*grids)
+
+    def get_fft(self, axes=None, direction='object', padding=None, shape=None, comm=None):
```

Comment:
It is not always the same result when we use padding, which we need for dealiasing.
When doing 3/2 dealiasing in RBC with Nx by Nz, we first do the inverse FFT along x, which results in a 3/2Nx by Nz shape, which we pad with zeros to 3/2Nx by 3/2Nz shape before the dct along the other axis, which we do by a then unpadded larger transform. So we need two FFT objects here. The reason we don't let mpi4py-fft do the padding in both directions is the weird counting of wavenumbers in FFT. We have to pad in a different place with the DCT (at the end) rather than the middle as in FFT. This did drive me mad.

Reply:
I was speaking about the test if MPI or not ... furthermore, it's hard to understand why we need `self.comm` and the `comm` attribute later, and why `self.comm` can be `None` and then replaced by the `comm` argument. I don't see how it relates to padding btw